python/mobitex_deframer.py

```diff
@@ -0,0 +1,366 @@
+        # 1st byte, bits 5-7
+        "message_type": (control_bytes[0] >> 5) & 0x07,
+        # 2nd byte, bit 0
+        "baud_bit": control_bytes[1] & 0x01,
+        # 2nd byte,bit 1
+        "ack_bit": (control_bytes[1] >> 1) & 0x01,
+        # 2nd byte,bits 2-3
+        "sub_address": (control_bytes[1] >> 2) & 0x03,
+        # 2nd byte,bits 4-7
+        "address": (control_bytes[1] >> 4) & 0x0F,
+    }
+
+    if (
+        status0 == Status.ERROR_UNCORRECTABLE or
+        status1 == Status.ERROR_UNCORRECTABLE
+    ):
+        raise ValueError("Control bytes FEC failed.")
+
+    # Count bit errors
+    bit_errors = 0
+    if status0 == Status.ERROR_CORRECTED:
+        bit_errors += 1
+    if status1 == Status.ERROR_CORRECTED:
+        bit_errors += 1
+
+    return result, bit_errors, control_bytes, control_byte_crc
+
+
+def decode_unknown_callsign(
+    callsign_bytes: bytes,
+    crc_bytes: bytes,
+    max_bit_flips: int,
+    callsign_ref: bytes = None,
+) -> tuple[bytes, int]:
+    """
+    Attempt to error-correct a callsign by flipping bits until CRC matches.
+
+    Args:
+        callsign_bytes: The callsign bytes to decode
+        crc_bytes: The CRC bytes to verify against
+        max_bit_flips: Maximum number of bits to flip during error correction
+        callsign_ref: Unused argument
+
+    Returns:
+        tuple containing:
+            - corrected callsign bytes
+            - number of bit errors corrected
+
+    Raises:
+        ValueError: If no valid callsign could be found
+                    within max_bit_flips bit flips
+    """
+    if max_bit_flips <= 0:
+        # Do not attempt error correction
+        computed_crc = crc16_ccitt_zero(bytes(callsign_bytes))
+        computed_crc_bytes = computed_crc.to_bytes(length=2)
+
+        if computed_crc_bytes == crc_bytes:
+            return callsign_bytes, 0
+        else:
+            if max_bit_flips == 0:
+                raise ValueError('Callsign CRC check failed.')
+            else:
+                # If max_bit_flips is negative, just accept invalid CRC
+                return callsign_bytes, -1
+
+    # Bit position 0... (N*8); -1 indicates "no flip"
+    all_bytes = callsign_bytes + crc_bytes
+    flip_positions = range(-1, len(all_bytes) * 8)
+
+    for flips in itertools.combinations(flip_positions, max_bit_flips):
+        # Create a modified copy with some bits flipped
+        mod_bytes = bytearray(all_bytes)
+        for i in flips:
+            if i == -1:
+                # No flip
+                continue
+            i_byte = i // 8
+            i_bit = i % 8
+
+            # Flip one bit
+            mod_bytes[i_byte] = mod_bytes[i_byte] ^ (1 << i_bit)
+
+        # Split back into callsign and CRC
+        mod_callsign = mod_bytes[: len(callsign_bytes)]
+        mod_crc = mod_bytes[len(callsign_bytes):]
+
+        # Check if modified callsign matches its CRC
+        computed_crc_bytes = crc16_ccitt_zero(mod_callsign).to_bytes(length=2)
```

Comment:
According to [the docs](https://docs.python.org/3/library/stdtypes.html#int.to_bytes), the default value for the `byteorder` argument of `to_bytes()` was added in Python 3.11, so it looks like this will need an explicit `byteorder='big'` to work with Python < 3.11.